metricbeat/helper/prometheus/hash.go

```diff
@@ -0,0 +1,72 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package prometheus
+
+import (
+	"bytes"
+	"sort"
+	"strconv"
+	"sync"
+
+	"github.com/cespare/xxhash"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+const sep = '\xff'
+
+var byteBuffer = sync.Pool{
+	New: func() interface{} {
+		return bytes.NewBuffer(nil)
+	},
+}
+
+type label struct {
+	key   string
+	value string
+}
+
+type labels []label
+
+func (ls labels) Len() int           { return len(ls) }
+func (ls labels) Swap(i, j int)      { ls[i], ls[j] = ls[j], ls[i] }
+func (ls labels) Less(i, j int) bool { return ls[i].key < ls[j].key }
+
+// LabelHash hashes the labels map and returns a string
+func LabelHash(labelMap common.MapStr) string {
+	ls := make(labels, len(labelMap))
+	b := byteBuffer.Get().(*bytes.Buffer)
+	b.Reset()
+
+	for k, v := range labelMap {
+		val, _ := v.(string)
+		ls = append(ls, label{k, val})
+	}
+
+	sort.Sort(ls)
+
+	for _, label := range ls {
+		b.WriteString(label.key)
+		b.WriteByte(sep)
+		b.WriteString(label.value)
+		b.WriteByte(sep)
+	}
+	hash := xxhash.Sum64(b.Bytes())
```

Comment:
How does writing to the temporary buffer compare to writing directly to a digest? As far as i can tell, xxhash lets you do something like this:

```
digest := xxhash.New()
for _, label := range ls {
    digest.WriteString(...)
    digest.WriteString(...)
}
```

Reply:
the use of separators ensures that ["ab/", "c"]  and ["a", "b/c"] dont hash to the same guy. hence why we use a byte buffer and add separators in between all key value pairs.

Reply:
This what happens if the byte buffer implementation is done with a string separator instead of a byte separator.

```
BenchmarkLabelHashWithFlatten-12       	  522225	      2098 ns/op	     800 B/op	      19 allocs/op
BenchmarkLabelHashWithOutFlatten-12    	 1253204	       946 ns/op	     544 B/op	       4 allocs/op
BenchmarkMapStrString-12               	  545319	      2259 ns/op	    1184 B/op	      24 allocs/op
PASS
```

This is what happens if we remove byte buffers and use the xxHash's internal `WriteString`.
```
goos: darwin
goarch: amd64
pkg: github.com/elastic/beats/metricbeat/helper/prometheus
BenchmarkLabelHashWithFlatten-12       	  554762	      2181 ns/op	     800 B/op	      19 allocs/op
BenchmarkLabelHashWithOutFlatten-12    	 1000000	      1093 ns/op	     544 B/op	       4 allocs/op
BenchmarkMapStrString-12               	  516709	      2187 ns/op	    1184 B/op	      24 allocs/op
PASS
```

The byte buffer seeems to be the faster of both.

Reply:
OK, thanks!